File: app/src/main/java/wangdaye/com/geometricweather/daily/adapter/DailyWeatherAdapter.java
package wangdaye.com.geometricweather.daily.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;
import java.util.TimeZone;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.options.unit.DurationUnit;
import wangdaye.com.geometricweather.common.basic.models.options.unit.PrecipitationUnit;
import wangdaye.com.geometricweather.common.basic.models.options.unit.ProbabilityUnit;
import wangdaye.com.geometricweather.common.basic.models.options.unit.TemperatureUnit;
import wangdaye.com.geometricweather.common.basic.models.weather.Daily;
import wangdaye.com.geometricweather.common.basic.models.weather.HalfDay;
import wangdaye.com.geometricweather.common.basic.models.weather.Precipitation;
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationDuration;
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationProbability;
import wangdaye.com.geometricweather.common.basic.models.weather.Temperature;
import wangdaye.com.geometricweather.daily.adapter.holder.AirQualityHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.AstroHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.LargeTitleHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.LineHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.MarginHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.OverviewHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.PollenHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.TitleHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.UVHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.ValueHolder;
import wangdaye.com.geometricweather.daily.adapter.holder.WindHolder;
import wangdaye.com.geometricweather.daily.adapter.model.DailyAirQuality;
import wangdaye.com.geometricweather.daily.adapter.model.DailyAstro;
import wangdaye.com.geometricweather.daily.adapter.model.DailyPollen;
import wangdaye.com.geometricweather.daily.adapter.model.DailyUV;
import wangdaye.com.geometricweather.daily.adapter.model.DailyWind;
import wangdaye.com.geometricweather.daily.adapter.model.LargeTitle;
import wangdaye.com.geometricweather.daily.adapter.model.Line;
import wangdaye.com.geometricweather.daily.adapter.model.Margin;
import wangdaye.com.geometricweather.daily.adapter.model.Overview;
import wangdaye.com.geometricweather.daily.adapter.model.Title;
import wangdaye.com.geometricweather.daily.adapter.model.Value;
import wangdaye.com.geometricweather.databinding.ItemWeatherDailyPollenBinding;
import wangdaye.com.geometricweather.settings.SettingsManager;

public class DailyWeatherAdapter extends RecyclerView.Adapter<DailyWeatherAdapter.ViewHolder> {

    private final List<ViewModel> mModelList;
    private final int mSpanCount;

    public interface ViewModel {
        int getCode();
    }

    public static abstract class ViewHolder extends RecyclerView.ViewHolder {

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public abstract void onBindView(ViewModel model, int position);
    }

    public GridLayoutManager.SpanSizeLookup spanSizeLookup = new GridLayoutManager.SpanSizeLookup() {
        @Override
        public int getSpanSize(int position) {
            int viewType = getItemViewType(position);
            if (Value.isCode(viewType)) {
                return 1;
            } else {
                return mSpanCount;
            }
        }
    };

    public DailyWeatherAdapter(Context context, TimeZone timeZone, Daily daily, int spanCount) {
        // model list.
        mModelList = new ArrayList<>();

        mModelList.add(new LargeTitle(context.getString(R.string.daytime)));
        mModelList.add(new Overview(daily.getDay(), true));
        mModelList.add(new DailyWind(daily.getDay().getWind()));
        mModelList.addAll(getHalfDayOptionalModelList(context, daily.getDay()));

        mModelList.add(new Line());
        mModelList.add(new LargeTitle(context.getString(R.string.nighttime)));
        mModelList.add(new Overview(daily.getNight(), false));
        mModelList.add(new DailyWind(daily.getNight().getWind()));
        mModelList.addAll(getHalfDayOptionalModelList(context, daily.getNight()));

        mModelList.add(new Line());
        mModelList.add(new LargeTitle(context.getString(R.string.life_details)));
        mModelList.add(new DailyAstro(timeZone, daily.getSun(), daily.getMoon(), daily.getMoonPhase()));
        if (daily.getAirQuality().isValid()) {
            mModelList.add(new Title(R.drawable.weather_haze_mini_xml, context.getString(R.string.air_quality)));
            mModelList.add(new DailyAirQuality(daily.getAirQuality()));
        }
        if (daily.getPollen().isValid()) {
            mModelList.add(new Title(R.drawable.ic_flower, context.getString(R.string.allergen)));
            mModelList.add(new DailyPollen(daily.getPollen()));
        }
        if (daily.getUV().isValid()) {
            mModelList.add(new Title(R.drawable.ic_uv, context.getString(R.string.uv_index)));
            mModelList.add(new DailyUV(daily.getUV()));
        }
        if (daily.getHoursOfSun() != null) {
            mModelList.add(new Line());
            mModelList.add(new Value(
                    context.getString(R.string.hours_of_sun),
                    DurationUnit.H.getValueText(context, daily.getHoursOfSun())
            ));
        }
        mModelList.add(new Margin());

        // span count.
        mSpanCount = spanCount;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (LargeTitle.isCode(viewType)) {
            return new LargeTitleHolder(parent);
        } else if (Overview.isCode(viewType)) {
            return new OverviewHolder(parent);
        } else if (Line.isCode(viewType)) {
            return new LineHolder(parent);
        } else if (Margin.isCode(viewType)) {
            return new MarginHolder(parent);
        } else if (Value.isCode(viewType)) {
            return new ValueHolder(parent);
        } else if (Title.isCode(viewType)) {
            return new TitleHolder(parent);
        } else if (DailyAirQuality.isCode(viewType)) {
            return new AirQualityHolder(parent);
        } else if (DailyAstro.isCode(viewType)) {
            return new AstroHolder(parent);
        } else if (DailyPollen.isCode(viewType)) {
            return new PollenHolder(
                    ItemWeatherDailyPollenBinding.inflate(
                            LayoutInflater.from(parent.getContext())
                    )
            );
        } else if (DailyUV.isCode(viewType)) {
            return new UVHolder(parent);
        } else if (DailyWind.isCode(viewType)) {
            return new WindHolder(parent);
        }
        throw new RuntimeException("Invalid viewType.");
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.onBindView(mModelList.get(position), position);
    }

    @Override
    public int getItemViewType(int position) {
        return mModelList.get(position).getCode();
    }

    @Override
    public int getItemCount() {
        return mModelList.size();
    }

    private List<ViewModel> getHalfDayOptionalModelList(Context context, HalfDay halfDay) {
        List<ViewModel> list = new ArrayList<>();
        // temperature.
        Temperature temperature = halfDay.getTemperature();
        TemperatureUnit temperatureUnit = SettingsManager.getInstance(context).getTemperatureUnit();
        if (temperature.getFeelsLikeTemperature() != null) {
            TemperatureUnit unit = SettingsManager.getInstance(context).getTemperatureUnit();
            int resId;
            if (unit == TemperatureUnit.C) {
                resId = R.drawable.ic_temperature_celsius;
            } else if (unit == TemperatureUnit.F) {
                resId = R.drawable.ic_temperature_fahrenheit;
            } else {
                resId = R.drawable.ic_temperature_kelvin;
            }
            list.add(new Title(resId, context.getString(R.string.temperature)));
            if (temperature.getRealFeelTemperature() != null) {
                list.add(new Value(
                        context.getString(R.string.real_feel_temperature),
                        temperatureUnit.getValueText(context, temperature.getRealFeelTemperature())
                ));
            }
            if (temperature.getRealFeelShaderTemperature() != null) {
                list.add(new Value(
                        context.getString(R.string.real_feel_shade_temperature),
                        temperatureUnit.getValueText(context, temperature.getRealFeelShaderTemperature())
                ));
            }
            if (temperature.getApparentTemperature() != null) {
                list.add(new Value(
                        context.getString(R.string.apparent_temperature),
                        temperatureUnit.getValueText(context, temperature.getApparentTemperature())
                ));
            }
            if (temperature.getWindChillTemperature() != null) {
                list.add(new Value(
                        context.getString(R.string.wind_chill_temperature),
                        temperatureUnit.getValueText(context, temperature.getWindChillTemperature())
                ));
            }
            if (temperature.getWetBulbTemperature() != null) {
                list.add(new Value(
                        context.getString(R.string.wet_bulb_temperature),
                        temperatureUnit.getValueText(context, temperature.getWetBulbTemperature())
                ));
            }
            if (temperature.getDegreeDayTemperature() != null) {
                list.add(new Value(
                        context.getString(R.string.degree_day_temperature),
                        temperatureUnit.getValueText(context, temperature.getDegreeDayTemperature())
                ));
            }
            list.add(new Margin());
        }

        // precipitation.
        Precipitation precipitation = halfDay.getPrecipitation();
        PrecipitationUnit precipitationUnit = SettingsManager.getInstance(context).getPrecipitationUnit();
        if (precipitation.getTotal() != null && precipitation.getTotal() > 0) {
            list.add(new Title(R.drawable.ic_water, context.getString(R.string.precipitation)));
            list.add(new Value(
                    context.getString(R.string.total),
                    precipitationUnit.getValueText(context, precipitation.getTotal())
            ));
            if (precipitation.getRain() != null && precipitation.getRain() > 0) {
                list.add(new Value(
                        context.getString(R.string.rain),
                        precipitationUnit.getValueText(context, precipitation.getRain())
                ));
            }
            if (precipitation.getSnow() != null && precipitation.getSnow() > 0) {
                list.add(new Value(
                        context.getString(R.string.snow),
                        precipitationUnit.getValueText(context, precipitation.getSnow())
                ));
            }
            if (precipitation.getIce() != null && precipitation.getIce() > 0) {
                list.add(new Value(
                        context.getString(R.string.ice),
                        precipitationUnit.getValueText(context, precipitation.getIce())
                ));
            }
            if (precipitation.getThunderstorm() != null && precipitation.getThunderstorm() > 0) {
                list.add(new Value(
                        context.getString(R.string.thunderstorm),
                        precipitationUnit.getValueText(context, precipitation.getThunderstorm())
                ));
            }
            list.add(new Margin());
        }

        // precipitation probability.
        PrecipitationProbability probability = halfDay.getPrecipitationProbability();
        if (probability.getTotal() != null && probability.getTotal() > 0) {
            list.add(new Title(R.drawable.ic_water_percent, context.getString(R.string.precipitation_probability)));
            list.add(new Value(
                    context.getString(R.string.total),
                    ProbabilityUnit.PERCENT.getValueText(context, (int) (float) probability.getTotal())
            ));
            if (probability.getRain() != null && probability.getRain() > 0) {
                list.add(new Value(
                        context.getString(R.string.rain),
                        ProbabilityUnit.PERCENT.getValueText(context, (int) (float) probability.getRain())
                ));
            }
            if (probability.getSnow() != null && probability.getSnow() > 0) {
                list.add(new Value(
                        context.getString(R.string.snow),
                        ProbabilityUnit.PERCENT.getValueText(context, (int) (float) probability.getSnow())
                ));
            }
            if (probability.getIce() != null && probability.getIce() > 0) {
                list.add(new Value(
                        context.getString(R.string.ice),
                        ProbabilityUnit.PERCENT.getValueText(context, (int) (float) probability.getIce())
                ));
            }
            if (probability.getThunderstorm() != null && probability.getThunderstorm() > 0) {
                list.add(new Value(
                        context.getString(R.string.thunderstorm),
                        ProbabilityUnit.PERCENT.getValueText(context, (int) (float) probability.getThunderstorm())
                ));
            }
            list.add(new Margin());
        }

        // precipitation duration.
        PrecipitationDuration duration = halfDay.getPrecipitationDuration();
        if (duration.getTotal() != null && duration.getTotal() > 0) {
            list.add(new Title(R.drawable.ic_time, context.getString(R.string.precipitation_duration)));
            list.add(new Value(
                    context.getString(R.string.total),
                    DurationUnit.H.getValueText(context, duration.getTotal())
            ));
            if (duration.getRain() != null && duration.getRain() > 0) {
                list.add(new Value(
                        context.getString(R.string.rain),
                        DurationUnit.H.getValueText(context, duration.getRain())
                ));
            }
            if (duration.getSnow() != null && duration.getSnow() > 0) {
                list.add(new Value(
                        context.getString(R.string.snow),
                        DurationUnit.H.getValueText(context, duration.getSnow())
                ));
            }
            if (duration.getIce() != null && duration.getIce() > 0) {
                list.add(new Value(
                        context.getString(R.string.ice),
                        DurationUnit.H.getValueText(context, duration.getIce())
                ));
            }
            if (duration.getThunderstorm() != null && duration.getThunderstorm() > 0) {
                list.add(new Value(
                        context.getString(R.string.thunderstorm),
                        DurationUnit.H.getValueText(context, duration.getThunderstorm())
                ));
            }
            list.add(new Margin());
        }
        return list;
    }
}

File: app/src/main/java/org/breezyweather/daily/adapter/DailyWeatherAdapter.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.daily.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.breezyweather.R
import org.breezyweather.common.basic.models.options.unit.DurationUnit
import org.breezyweather.common.basic.models.options.unit.ProbabilityUnit
import breezyweather.domain.weather.model.Daily
import breezyweather.domain.weather.model.HalfDay
import org.breezyweather.daily.adapter.holder.*
import org.breezyweather.daily.adapter.model.*
import org.breezyweather.databinding.ItemWeatherDailyPollenBinding
import org.breezyweather.domain.weather.model.isIndexValid
import org.breezyweather.settings.SettingsManager
import java.util.TimeZone

class DailyWeatherAdapter(context: Context, timeZone: TimeZone, daily: Daily, spanCount: Int) :
    RecyclerView.Adapter<DailyWeatherAdapter.ViewHolder>() {
    private val mModelList: MutableList<ViewModel>
    private val mSpanCount: Int = spanCount

    interface ViewModel {
        val code: Int
    }

    abstract class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        abstract fun onBindView(model: ViewModel, position: Int)
    }

    var spanSizeLookup: GridLayoutManager.SpanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
        override fun getSpanSize(position: Int): Int {
            val viewType = getItemViewType(position)
            return if (Value.isCode(viewType)) 1 else mSpanCount
        }
    }

    init {
        // model list.
        mModelList = ArrayList()
        daily.day?.let { day ->
            mModelList.add(LargeTitle(context.getString(R.string.daytime)))
            mModelList.add(Overview(day, true))
            day.wind?.let { wind ->
                if (wind.isValid) {
                    mModelList.add(DailyWind(wind))
                }
            }
            mModelList.addAll(getHalfDayOptionalModelList(context, day))
        }
        daily.night?.let { night ->
            mModelList.add(Line())
            mModelList.add(LargeTitle(context.getString(R.string.nighttime)))
            mModelList.add(Overview(night, false))
            night.wind?.let { wind ->
                if (wind.isValid) {
                    mModelList.add(DailyWind(wind))
                }
            }
            mModelList.addAll(getHalfDayOptionalModelList(context, night))
        }
        mModelList.add(Line())
        daily.airQuality?.let { airQuality ->
            if (airQuality.isIndexValid) {
                mModelList.add(
                    Title(
                        R.drawable.weather_haze_mini_xml,
                        context.getString(R.string.air_quality)
                    )
                )
                mModelList.add(DailyAirQuality(airQuality))
            }
        }
        daily.pollen?.let { pollen ->
            if (pollen.isIndexValid) {
                mModelList.add(
                    Title(
                        R.drawable.ic_allergy,
                        context.getString(if (pollen.isMoldValid) R.string.pollen_and_mold else R.string.pollen)
                    )
                )
                mModelList.add(DailyPollen(pollen))
            }
        }
        daily.uV?.let { uV ->
            if (uV.isValid) {
                mModelList.add(Title(R.drawable.ic_uv, context.getString(R.string.uv_index)))
                mModelList.add(DailyUV(uV))
            }
        }
        if (daily.sun?.isValid == true || daily.moon?.isValid == true || daily.moonPhase?.isValid == true) {
            mModelList.add(LargeTitle(context.getString(R.string.ephemeris)))
            mModelList.add(DailyAstro(timeZone, daily.sun, daily.moon, daily.moonPhase))
        }
        if (daily.degreeDay?.isValid == true || daily.hoursOfSun != null) {
            mModelList.add(Line())
            mModelList.add(LargeTitle(context.getString(R.string.details)))
            daily.degreeDay?.let { degreeDay ->
                if (degreeDay.isValid) {
                    val temperatureUnit = SettingsManager.getInstance(context).temperatureUnit
                    if ((degreeDay.heating ?: 0.0) > 0) {
                        mModelList.add(
                            ValueIcon(
                                context.getString(R.string.temperature_degree_day_heating),
                                temperatureUnit.getDegreeDayValueText(
                                    context,
                                    degreeDay.heating!!
                                ),
                                R.drawable.ic_mode_heat
                            )
                        )
                    } else if ((degreeDay.cooling ?: 0.0) > 0) {
                        mModelList.add(
                            ValueIcon(
                                context.getString(R.string.temperature_degree_day_cooling),
                                temperatureUnit.getDegreeDayValueText(
                                    context,
                                    degreeDay.cooling!!
                                ),
                                R.drawable.ic_mode_cool
                            )
                        )
                    }
                }
            }
            daily.hoursOfSun?.let { hoursOfSun ->
                mModelList.add(
                    ValueIcon(
                        context.getString(R.string.hours_of_sun),
                        DurationUnit.H.getValueText(context, hoursOfSun),
                        R.drawable.ic_hours_of_sun

                    )
                )
            }
        }
        mModelList.add(Margin())
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        if (LargeTitle.isCode(viewType)) {
            return LargeTitleHolder(parent)
        } else if (Overview.isCode(viewType)) {
            return OverviewHolder(parent)
        } else if (Line.isCode(viewType)) {
            return LineHolder(parent)
        } else if (Margin.isCode(viewType)) {
            return MarginHolder(parent)
        } else if (Value.isCode(viewType)) {
            return ValueHolder(parent)
        } else if (ValueIcon.isCode(viewType)) {
            return ValueIconHolder(parent)
        } else if (Title.isCode(viewType)) {
            return TitleHolder(parent)
        } else if (ValueIcon.isCode(viewType)) {
            return ValueIconHolder(parent)
        } else if (DailyAirQuality.isCode(viewType)) {
            return AirQualityHolder(parent)
        } else if (DailyAstro.isCode(viewType)) {
            return AstroHolder(parent)
        } else if (DailyPollen.isCode(viewType)) {
            return PollenHolder(
                ItemWeatherDailyPollenBinding.inflate(
                    LayoutInflater.from(parent.context)
                )
            )
        } else if (DailyUV.isCode(viewType)) {
            return UVHolder(parent)
        } else if (DailyWind.isCode(viewType)) {
            return WindHolder(parent)
        }
        throw RuntimeException("Invalid viewType.")
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.onBindView(mModelList[position], position)
    }

    override fun getItemViewType(position: Int) = mModelList[position].code

    override fun getItemCount() = mModelList.size

    private fun getHalfDayOptionalModelList(context: Context, halfDay: HalfDay): List<ViewModel> {
        val list: MutableList<ViewModel> = ArrayList()
        // temperature.
        val temperature = halfDay.temperature
        val temperatureUnit = SettingsManager.getInstance(context).temperatureUnit
        if (temperature?.feelsLikeTemperature != null) {
            list.add(Title(R.drawable.ic_device_thermostat, context.getString(R.string.temperature)))
            temperature.realFeelTemperature?.let {
                list.add(
                    Value(
                        context.getString(R.string.temperature_real_feel),
                        temperatureUnit.getValueText(context, it)
                    )
                )
            }
            temperature.realFeelShaderTemperature?.let {
                list.add(
                    Value(
                        context.getString(R.string.temperature_real_feel_shade),
                        temperatureUnit.getValueText(context, it)
                    )
                )
            }
            temperature.apparentTemperature?.let {
                list.add(
                    Value(
                        context.getString(R.string.temperature_apparent),
                        temperatureUnit.getValueText(context, it)
                    )
                )
            }
            temperature.windChillTemperature?.let {
                list.add(
                    Value(
                        context.getString(R.string.temperature_wind_chill),
                        temperatureUnit.getValueText(context, it)
                    )
                )
            }
            temperature.wetBulbTemperature?.let {
                list.add(
                    Value(
                        context.getString(R.string.temperature_wet_bulb),
                        temperatureUnit.getValueText(context, it)
                    )
                )
            }
            list.add(Margin())
        }

        // precipitation.
        val precipitation = halfDay.precipitation
        val precipitationUnit = SettingsManager.getInstance(context).precipitationUnit
        if ((precipitation?.total ?: 0.0) > 0) {
            list.add(Title(R.drawable.ic_water, context.getString(R.string.precipitation)))
            list.add(
                Value(
                    context.getString(R.string.precipitation_total),
                    precipitationUnit.getValueText(context, precipitation!!.total!!)
                )
            )
            if ((precipitation.rain ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_rain),
                        precipitationUnit.getValueText(context, precipitation.rain!!)
                    )
                )
            }
            if ((precipitation.snow ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_snow),
                        precipitationUnit.getValueText(context, precipitation.snow!!)
                    )
                )
            }
            if ((precipitation.ice ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_ice),
                        precipitationUnit.getValueText(context, precipitation.ice!!)
                    )
                )
            }
            if ((precipitation.thunderstorm ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_thunderstorm),
                        precipitationUnit.getValueText(context, precipitation.thunderstorm!!)
                    )
                )
            }
            list.add(Margin())
        }

        // precipitation probability.
        val probability = halfDay.precipitationProbability
        if ((probability?.total ?: 0.0) > 0) {
            list.add(Title(R.drawable.ic_water_percent, context.getString(R.string.precipitation_probability)))
            list.add(
                Value(
                    context.getString(R.string.precipitation_total),
                    ProbabilityUnit.PERCENT.getValueText(context, probability!!.total!!.toInt())
                )
            )
            if ((probability.rain ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_rain),
                        ProbabilityUnit.PERCENT.getValueText(context, probability.rain!!.toInt())
                    )
                )
            }
            if ((probability.snow ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_snow),
                        ProbabilityUnit.PERCENT.getValueText(context, probability.snow!!.toInt())
                    )
                )
            }
            if ((probability.ice ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_ice),
                        ProbabilityUnit.PERCENT.getValueText(context, probability.ice!!.toInt())
                    )
                )
            }
            if ((probability.thunderstorm ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_thunderstorm),
                        ProbabilityUnit.PERCENT.getValueText(context, probability.thunderstorm!!.toInt())
                    )
                )
            }
            list.add(Margin())
        }

        // precipitation duration.
        val duration = halfDay.precipitationDuration
        if ((duration?.total ?: 0.0) > 0) {
            list.add(Title(R.drawable.ic_time, context.getString(R.string.precipitation_duration)))
            list.add(
                Value(
                    context.getString(R.string.precipitation_total),
                    DurationUnit.H.getValueText(context, duration!!.total!!)
                )
            )
            if ((duration.rain ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_rain),
                        DurationUnit.H.getValueText(context, duration.rain!!)
                    )
                )
            }
            if ((duration.snow ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_snow),
                        DurationUnit.H.getValueText(context, duration.snow!!)
                    )
                )
            }
            if ((duration.ice ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_ice),
                        DurationUnit.H.getValueText(context, duration.ice!!)
                    )
                )
            }
            if ((duration.thunderstorm ?: 0.0) > 0) {
                list.add(
                    Value(
                        context.getString(R.string.precipitation_thunderstorm),
                        DurationUnit.H.getValueText(context, duration.thunderstorm!!)
                    )
                )
            }
            list.add(Margin())
        }
        return list
    }
}
